File: tests/test_resend_send.mjs
// Outbound mail through Resend, exercised against the shipped module.
//
// Graph did two things for us that direct sending does not: it threaded replies
// (/messages/{id}/reply) and it put the mailbox's display name on the envelope.
// Both are now ours to get right, and both are invisible when wrong — a reply
// that starts a new thread still arrives, and so does one signed with a bare
// address. These pin them down.
import assert from "node:assert/strict";

process.env.RESEND_API_KEY = "test-key";
const { sendViaResend, resendConfigured } = await import("../apps/provisioning-service/src/clients/resend.ts");

const results = [];
function test(name, fn) {
  return fn().then(
    () => results.push(`ok   ${name}`),
    (e) => results.push(`FAIL ${name}: ${e.message}`),
  );
}

/** Capture the request Resend would receive, without making one. */
let lastRequest = null;
function stubFetch({ ok = true, status = 200, body = '{"id":"re_123"}' } = {}) {
  globalThis.fetch = async (url, init) => {
    lastRequest = { url, init, payload: JSON.parse(init.body) };
    return {
      ok,
      status,
      text: async () => body,
    };
  };
}

const BASE = {
  from: "agent@agents.example.net",
  to: ["buyer@example.com"],
  subject: "Re: September expenses",
  body: "<p>Here are the totals.</p>",
};

await test("the message is posted to resend with the api key", async () => {
  stubFetch();
  const r = await sendViaResend(BASE);
  assert.equal(r.ok, true);
  assert.equal(r.id, "re_123");
  assert.equal(lastRequest.url, "https://api.resend.com/emails");
  assert.equal(lastRequest.init.headers.Authorization, "Bearer test-key");
});

await test("a display name is carried so the agent writes under its own name", async () => {
  stubFetch();
  await sendViaResend({ ...BASE, fromName: "Data Analyst Test" });
  assert.equal(lastRequest.payload.from, '"Data Analyst Test" <agent@agents.example.net>');
});

await test("a name with a quote in it cannot rewrite the header", async () => {
  // The agent name is set by the buyer. A bare quote would close the
  // quoted-string early and let the rest be read as header syntax.
  stubFetch();
  await sendViaResend({ ...BASE, fromName: 'Evil" <attacker@elsewhere.test>, "x' });
  const from = lastRequest.payload.from;
  // The address must still be ours, and the name must remain a single quoted
  // string - an address inside the quotes is inert, one outside them is not.
  assert.ok(from.endsWith(" <agent@agents.example.net>"), `got: ${from}`);
  const name = from.slice(0, -" <agent@agents.example.net>".length);
  assert.ok(name.startsWith('"') && name.endsWith('"'), `name not quoted: ${name}`);
  assert.equal(name.slice(1, -1).includes('"'), false, `quote survived: ${name}`);
});

await test("a newline in the name cannot inject a header", async () => {
  stubFetch();
  await sendViaResend({ ...BASE, fromName: "Agent\r\nBcc: someone@elsewhere.test" });
  assert.ok(!/[\r\n]/.test(lastRequest.payload.from));
});

await test("no display name sends as the bare address", async () => {
  stubFetch();
  await sendViaResend(BASE);
  assert.equal(lastRequest.payload.from, "agent@agents.example.net");
});

await test("a reply carries in-reply-to so it stays in the thread", async () => {
  stubFetch();
  await sendViaResend({ ...BASE, inReplyTo: "<abc@mail.example.com>" });
  assert.equal(lastRequest.payload.headers["In-Reply-To"], "<abc@mail.example.com>");
  assert.equal(lastRequest.payload.headers["References"], "<abc@mail.example.com>");
});

await test("an existing references chain is extended, not replaced", async () => {
  stubFetch();
  await sendViaResend({
    ...BASE,
    inReplyTo: "<second@mail.example.com>",
    references: "<first@mail.example.com>",
  });
  assert.equal(
    lastRequest.payload.headers["References"],
    "<first@mail.example.com> <second@mail.example.com>",
  );
});

await test("a message that answers nothing sends no threading headers", async () => {
  // A fabricated In-Reply-To threads the message nowhere and looks forged.
  stubFetch();
  await sendViaResend(BASE);
  assert.equal(lastRequest.payload.headers, undefined);
});

await test("html and text bodies use their own field", async () => {
  stubFetch();
  await sendViaResend({ ...BASE, bodyType: "text", body: "plain words" });
  assert.equal(lastRequest.payload.text, "plain words");
  assert.equal(lastRequest.payload.html, undefined);

  stubFetch();
  await sendViaResend(BASE);
  assert.equal(lastRequest.payload.html, "<p>Here are the totals.</p>");
  assert.equal(lastRequest.payload.text, undefined);
});

await test("attachments are renamed to what resend expects", async () => {
  stubFetch();
  await sendViaResend({
    ...BASE,
    attachments: [{ name: "totals.xlsx", contentType: "application/vnd.ms-excel", content_base64: "AAA=" }],
  });
  const [a] = lastRequest.payload.attachments;
  assert.equal(a.filename, "totals.xlsx");
  assert.equal(a.content, "AAA=");
  assert.equal(a.content_type, "application/vnd.ms-excel");
});

await test("a rejection is reported rather than swallowed", async () => {
  // The caller falls back to Graph on a false, so this must not throw and must
  // not claim success - the whole point of the move is to stop reporting
  // "sent" for mail that was not.
  stubFetch({ ok: false, status: 422, body: '{"message":"domain not verified"}' });
  const r = await sendViaResend(BASE);
  assert.equal(r.ok, false);
  assert.equal(r.status, 422);
  assert.match(r.error, /not verified/);
});

await test("a network failure is reported rather than thrown", async () => {
  globalThis.fetch = async () => {
    throw new Error("socket hang up");
  };
  const r = await sendViaResend(BASE);
  assert.equal(r.ok, false);
  assert.match(r.error, /socket hang up/);
});

await test("a 2xx with an unreadable body still counts as sent", async () => {
  stubFetch({ body: "not json" });
  const r = await sendViaResend(BASE);
  assert.equal(r.ok, true);
  assert.equal(r.id, undefined);
});

assert.equal(resendConfigured(), true);

console.log(results.join("\n"));
if (results.some((r) => r.startsWith("FAIL"))) process.exit(1);

File: tests/test_resend_send.py
"""Outbound mail moves to Resend; the mailbox it is sent from does not.

Gmail began refusing this domain's mail outright on 2026-09-20 — 550 5.7.1,
"likely unsolicited mail" — after two days of test traffic. Authentication was
correct the whole time, so the verdict was about the sender: Exchange Online's
shared outbound IPs are not a transactional sender and lend a young domain no
reputation. Sending moves to a service built for it.

The behaviour is tested in test_resend_send.mjs so it runs the shipped module;
this file runs it under pytest (which is what CI runs) and checks the two send
paths go through it, keep Graph as a fallback, and preserve the two things Graph
was quietly doing for us: threading and the sender's display name.
"""
import shutil
import subprocess
from pathlib import Path

import pytest

ROOT = Path(__file__).resolve().parents[1]
SERVER = ROOT / "apps" / "provisioning-service" / "src" / "server.ts"
CLIENT = ROOT / "apps" / "provisioning-service" / "src" / "clients" / "resend.ts"
SCRIPT = Path(__file__).resolve().parent / "test_resend_send.mjs"


@pytest.mark.skipif(shutil.which("node") is None, reason="node is needed")
def test_the_sending_behaves():
    r = subprocess.run(
        ["node", "--experimental-strip-types", str(SCRIPT)],
        capture_output=True, text=True, cwd=SCRIPT.parent.parent,
    )
    print(r.stdout or r.stderr)
    assert r.returncode == 0, r.stdout + r.stderr
    assert r.stdout.count("ok   ") >= 12, r.stdout


@pytest.mark.parametrize("endpoint", ["platform-send", "outlook-send"])
def test_both_send_paths_try_resend_first(endpoint):
    src = SERVER.read_text(encoding="utf-8")
    start = src.index(f'req.url === "/internal/{endpoint}"')
    handler = src[start:start + 6000]
    assert "resendConfigured()" in handler, f"{endpoint} never tries Resend"
    assert "sendViaResend" in handler


@pytest.mark.parametrize("endpoint", ["platform-send", "outlook-send"])
def test_a_resend_failure_falls_back_to_graph(endpoint):
    # A missing or rejected key must not be the reason a buyer never hears from
    # their agent. Graph stays wired up underneath.
    src = SERVER.read_text(encoding="utf-8")
    start = src.index(f'req.url === "/internal/{endpoint}"')
    handler = src[start:start + 6000]
    assert "falling back to Graph" in handler
    graph_at = handler.index("graph.microsoft.com")
    resend_at = handler.index("sendViaResend")
    assert resend_at < graph_at, "Graph should be the fallback, not the first choice"


def test_a_reply_keeps_its_thread():
    # Graph's /messages/{id}/reply threaded replies for us. Sending directly does
    # not, and an agent whose answer arrives as a new conversation has lost the
    # request it was answering.
    src = SERVER.read_text(encoding="utf-8")
    assert "threadHeadersFor" in src
    assert "internetMessageId" in src
    start = src.index('req.url === "/internal/outlook-send"')
    handler = src[start:start + 6000]
    assert "replyToMessageId" in handler and "threadHeadersFor" in handler


def test_the_agent_still_writes_under_its_own_name():
    # Graph put the mailbox's display name on the envelope. Losing it would have
    # every agent suddenly writing to its buyer as a raw address.
    src = SERVER.read_text(encoding="utf-8")
    assert "mailboxDisplayName" in src
    for endpoint in ("platform-send", "outlook-send"):
        start = src.index(f'req.url === "/internal/{endpoint}"')
        assert "mailboxDisplayName" in src[start:start + 6000], endpoint


def test_neither_lookup_can_fail_a_send():
    # Both are conveniences read from Graph. Neither is worth losing a message
    # over, so both swallow their errors and return nothing.
    src = SERVER.read_text(encoding="utf-8")
    for fn in ("mailboxDisplayName", "threadHeadersFor"):
        start = src.index(f"async function {fn}")
        body = src[start:src.index("\n}\n", start)]
        assert "try {" in body and "catch" in body, f"{fn} can throw into a send"


def test_the_client_stands_alone():
    # Imported directly by its test, so it must not drag the service config in.
    src = CLIENT.read_text(encoding="utf-8")
    assert 'from "../config.js"' not in src
    assert "process.env.RESEND_API_KEY" in src


def test_a_reply_is_titled_after_what_it_answers():
    # Graph's reply endpoint built "Re: <subject>" for us, and the adapter sends
    # no subject at all on a reply. Falling back to a bare "Re:" shipped on
    # 2026-09-20 and the reply opened its own conversation in Gmail, which groups
    # on the subject as well as the References chain.
    src = SERVER.read_text(encoding="utf-8")
    assert "function replySubjectFor" in src
    start = src.index('req.url === "/internal/outlook-send"')
    handler = src[start:start + 6000]
    assert "replySubjectFor(thread.subject)" in handler
    assert '|| "Re:";' not in handler, "a reply must not fall back to a bare Re:"


def test_the_subject_of_the_answered_message_is_fetched():
    src = SERVER.read_text(encoding="utf-8")
    start = src.index("async function threadHeadersFor")
    body = src[start:src.index("\n}\n", start)]
    assert "subject" in body, "the subject must come back with the threading headers"


def test_re_is_not_stacked():
    # "Re: Re: Re:" is what happens when a reply to a reply re-prefixes blindly.
    src = SERVER.read_text(encoding="utf-8")
    start = src.index("function replySubjectFor")
    body = src[start:src.index("\n}\n", start)]
    assert "/^re:/i" in body


def test_a_reply_that_cannot_thread_says_so():
    # Silence here is what made this cost a whole test run to find: the send
    # succeeded, the mail arrived, and only the threading was wrong.
    src = SERVER.read_text(encoding="utf-8")
    assert "this will start a new thread" in src
